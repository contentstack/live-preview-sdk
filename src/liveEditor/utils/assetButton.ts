/**
 * Removes all existing replace asset buttons from the provided visual editor wrapper element.
 * @param visualEditorWrapper - The visual editor wrapper element to remove replace asset buttons from.
 */
export function removeReplaceAssetButton(
    visualEditorWrapper: HTMLDivElement | null
): void {
    if (!visualEditorWrapper) {
        return;
    }

    const existingReplaceButtons = visualEditorWrapper.getElementsByClassName(
        "visual-editor__replace-button"
    );

    for (const existingReplaceButton of Array.from(existingReplaceButtons)) {
        existingReplaceButton.remove();
    }
}

/**
 * Generates a replace asset button element that can be appended to the DOM.
 * @param targetElement - The target element to attach the button to.
 * @param onClickCallback - The callback function to execute when the button is clicked.
 * @returns The generated HTMLButtonElement.
 */
export function generateReplaceAssetButton(
    targetElement: Element,
    onClickCallback: (event: MouseEvent) => void
): HTMLButtonElement {
    const dimension = targetElement.getBoundingClientRect();
    const replaceButton = document.createElement("button");

    replaceButton.addEventListener("click", onClickCallback);
    replaceButton.classList.add("visual-editor__replace-button");
    replaceButton.setAttribute("data-testid", "visual-editor-replace-asset");
    replaceButton.innerHTML = `Replace Asset`;
    replaceButton.style.top = `${dimension.bottom + window.scrollY - 30}px`;
    replaceButton.style.right = `${window.innerWidth - dimension.right}px`;

    return replaceButton;
}
