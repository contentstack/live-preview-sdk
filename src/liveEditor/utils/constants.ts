export const numericInputRegex = /^-?\d*(\.\d*)?([eE][-+]?\d*)?$/;

export const LIVE_EDITOR_FIELD_TYPE_ATTRIBUTE_KEY = "data-cslp-field-type";

export const LIVE_EDITOR_CHANNEL_ID = "live-editor";

export const LIVE_PREVIEW_OUTLINE_WIDTH_IN_PX = 2;
