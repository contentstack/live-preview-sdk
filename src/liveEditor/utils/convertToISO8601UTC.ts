export function convertToISO8601UTC(
    dateString: string,
    timeString: string
): string {
    // Extract the time and offset
    const time = timeString.slice(0, 8); // '00:00:22'
    const offset = timeString.slice(8); // '+0530'

    // get the only date part of the given date
    const datePart = dateString.split("T")[0];

    // Combine date and time into a single string
    const dateTimeString = `${datePart} ${time}${offset}`;

    // Parse the dateTimeString as a Date object
    const date = new Date(dateTimeString);

    // Convert the date to ISO 8601 format in UTC
    const iso8601UTC = date.toISOString();

    return iso8601UTC;
}
