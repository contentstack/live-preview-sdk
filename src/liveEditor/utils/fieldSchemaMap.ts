import { has, isEqual } from "lodash";
import {
    ISchemaFieldMap,
    ISchemaIndividualFieldMap,
} from "./types/index.types";

import liveEditorPostMessage from "./liveEditorPostMessage";
import { LiveEditorPostMessageEvents } from "./types/postMessage.types";

/**
 * Represents a cache for field schemas. Field schemas are
 * used to easily get the field schema based on the field
 * Cslp.
 */
export class FieldSchemaMap {
    private static fieldSchema: {
        [contentTypeUid: string]: ISchemaIndividualFieldMap;
    } = {};

    /**
     * Retrieves the schema field map for a given content type and field Cslp.
     * @param contentTypeUid - The unique identifier of the content type.
     * @param fieldCslp - The Cslp of the field.
     * @returns The schema field map.
     */
    static async getFieldSchema(
        contentTypeUid: string,
        fieldCslp: string
    ): Promise<ISchemaFieldMap> {
        if (FieldSchemaMap.hasFieldSchema(contentTypeUid, fieldCslp)) {
            return Promise.resolve(
                FieldSchemaMap.fieldSchema[contentTypeUid][fieldCslp]
            );
        }

        const data = await liveEditorPostMessage?.send<{
            fieldSchemaMap: ISchemaIndividualFieldMap;
        }>(LiveEditorPostMessageEvents.GET_FIELD_SCHEMA, {
            contentTypeUid: contentTypeUid,
        });

        if (data?.fieldSchemaMap) {
            FieldSchemaMap.fieldSchema[contentTypeUid] = data.fieldSchemaMap;
        }

        return FieldSchemaMap.fieldSchema[contentTypeUid][fieldCslp];
    }

    static hasFieldSchema(contentTypeUid: string, fieldCslp: string): boolean {
        return has(FieldSchemaMap.fieldSchema, [contentTypeUid, fieldCslp]);
    }

    /**
     * Checks if two field schemas are equal.
     * @param firstFieldSchema - The first field schema to compare.
     * @param secondFieldSchema - The second field schema to compare.
     * @returns True if the field schemas are equal, false otherwise.
     */
    static areFieldSchemaEqual(
        firstFieldSchema: ISchemaFieldMap,
        secondFieldSchema: ISchemaFieldMap
    ): boolean {
        return isEqual(firstFieldSchema, secondFieldSchema);
    }

    /**
     * Sets the field schema for a given content type.
     * @param contentTypeUid The unique identifier of the content type.
     * @param fieldSchemaMap The map of individual field schemas.
     */
    static setFieldSchema(
        contentTypeUid: string,
        fieldSchemaMap: ISchemaIndividualFieldMap
    ): void {
        FieldSchemaMap.fieldSchema[contentTypeUid] = fieldSchemaMap;
    }

    /**
     * Clears the field schema cache.
     */
    static clear(): void {
        FieldSchemaMap.fieldSchema = {};
    }
}
