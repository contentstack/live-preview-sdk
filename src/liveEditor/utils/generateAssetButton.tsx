import { render } from "preact";
import ReplaceAssetButtonComponent from "../components/replaceAssetButton";

/**
 * Removes all existing replace asset buttons from the provided visual editor wrapper element.
 * @param visualEditorWrapper - The visual editor wrapper element to remove replace asset buttons from.
 */
export function removeReplaceAssetButton(
    visualEditorWrapper: HTMLDivElement | null
): void {
    if (!visualEditorWrapper) {
        return;
    }

    const existingReplaceButtons = visualEditorWrapper.getElementsByClassName(
        "visual-editor__replace-button"
    );

    for (const existingReplaceButton of Array.from(existingReplaceButtons)) {
        existingReplaceButton.remove();
    }
}

/**
 * Generates a replace asset button element that can be appended to the DOM.
 * @param targetElement - The target element to attach the button to.
 * @param onClickCallback - The callback function to execute when the button is clicked.
 * @returns The generated HTMLButtonElement.
 */
export function generateReplaceAssetButton(
    targetElement: Element,
    onClickCallback: (event: MouseEvent) => void
): HTMLButtonElement {
   
    const visualEditorWrapper = document.querySelector(".visual-editor__container");
    const wrapper = document.createDocumentFragment();
    render(<ReplaceAssetButtonComponent targetElement={targetElement} onClickCallback={onClickCallback}/>, wrapper);
    
    visualEditorWrapper?.appendChild(wrapper);

    const replaceButton = document.querySelector(".visual-editor__replace-button") as HTMLButtonElement;

    return replaceButton;
}
