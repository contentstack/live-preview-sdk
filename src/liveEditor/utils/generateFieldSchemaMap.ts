import mockData from "../ctmap";
import _ from "lodash";
import { ITraverseSchemaVisitor } from "./types/index.types";

export function generateFieldSchemaMap(ctUID: string): Record<string, any> {
    const pageCT = mockData[ctUID];
    const getFieldSchemaMap: ITraverseSchemaVisitor = {
        fieldMap: {},
        should_visit: (_fieldSchema, _path) => {
            return true;
        },
        visit: function (fieldSchema, path) {
            this.fieldMap[path] = fieldSchema;
            if (fieldSchema.data_type === "link") {
                //handle special key for special fields
                this.fieldMap[`${path}.title`] = fieldSchema;
                this.fieldMap[`${path}.url`] = fieldSchema;
            }
            if (fieldSchema.data_type === "file") {
                this.fieldMap[`${path}.url`] = fieldSchema;
            }
            if (fieldSchema.data_type === "blocks") {
                if (!fieldSchema.blocks) {
                    return;
                }
                fieldSchema.blocks.map((block: any) => {
                    this.fieldMap[`${path}.${block.uid}`] = {
                        ...block,
                        data_type: "block",
                        display_name: block.title,
                    };
                });
            }
        },
    };
    traverseSchema(pageCT.schema, [getFieldSchemaMap]);
    return getFieldSchemaMap.fieldMap;
}

function traverseSchema(
    schema: Record<string, unknown>,
    visitors: Array<ITraverseSchemaVisitor>
): void {
    function genPath(prefix: string, path: string) {
        return _.isEmpty(prefix) ? path : [prefix, path].join(".");
    }

    function traverse(fields: any, path: string) {
        path = path || "";
        for (const element of fields) {
            const field = element;
            const currPath = genPath(path, field.uid);

            visitors.forEach((visitor) => {
                if (visitor.should_visit(field, currPath)) {
                    visitor.visit(field, currPath);
                }
            });

            if (field.data_type === "group") traverse(field.schema, currPath);

            if (
                field.data_type === "global_field" &&
                _.isUndefined(field.schema) === false &&
                _.isEmpty(field.schema) === false
            )
                traverse(field.schema, currPath);
            if (field.data_type === "blocks") {
                field.blocks.forEach(function (block: any) {
                    if (block.schema)
                        traverse(block.schema, currPath + "." + block.uid);
                });
            }
            if (field.data_type === "experience_container") {
                field.variations.forEach(function (variation: any) {
                    if (variation.schema)
                        traverse(
                            variation.schema,
                            currPath + "." + variation.uid
                        );
                });
            }
        }
    }
    traverse(schema, "");
}
