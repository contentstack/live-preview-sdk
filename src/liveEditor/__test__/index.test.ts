import crypto from "crypto";

import { getFieldSchemaMap } from "../../__test__/data/fieldSchemaMap";
import { sleep } from "../../__test__/utils";
import { VisualEditor } from "../index";
import { FieldSchemaMap } from "../utils/fieldSchemaMap";

jest.mock("../utils/liveEditorPostMessage", () => {
    const { getAllContentTypes } = jest.requireActual(
        "../../__test__/data/contentType"
    );
    const contentTypes = getAllContentTypes();
    const liveEditorPostMessageActual = jest.requireActual(
        "../utils/liveEditorPostMessage"
    );

    return {
        __esModule: true,
        default: {
            ...liveEditorPostMessageActual.default,
            send: jest.fn().mockImplementation((eventName: string) => {
                if (eventName === "init")
                    return Promise.resolve({
                        contentTypes,
                    });
                return Promise.resolve();
            }),
        },
    };
});

Object.defineProperty(globalThis, "crypto", {
    value: {
        getRandomValues: (arr: Array<any>) => crypto.randomBytes(arr.length),
    },
});

global.ResizeObserver = jest.fn().mockImplementation(() => ({
    observe: jest.fn(),
    unobserve: jest.fn(),
    disconnect: jest.fn(),
}));

describe("Visual editor", () => {
    beforeAll(() => {
        FieldSchemaMap.setFieldSchema(
            "all_fields",
            getFieldSchemaMap().all_fields
        );
    });

    afterEach(() => {
        document.getElementsByTagName("html")[0].innerHTML = "";
    });

    afterAll(() => {
        FieldSchemaMap.clear();
    });

    test("should append a visual editor container to the DOM", () => {
        let visualEditorDOM = document.querySelector(
            ".visual-editor__container"
        );

        expect(visualEditorDOM).toBeNull();

        new VisualEditor();

        visualEditorDOM = document.querySelector(
            `[data-testid="visual-editor__container"]`
        );

        expect(visualEditorDOM).toMatchSnapshot();
    });

    describe("inline editing", () => {
        let h1Tag: HTMLHeadingElement;
        beforeEach(() => {
            h1Tag = document.createElement("h1");
            h1Tag.innerText = "Hello World";

            h1Tag.setAttribute(
                "data-cslp",
                "all_fields.blt58a50b4cebae75c5.en-us.modular_blocks.0.block.single_line"
            );
            document.body.appendChild(h1Tag);
        });

        test("should add overlay to DOM when clicked", async () => {
            new VisualEditor();

            await sleep(0);
            h1Tag.click();
            await sleep(0);

            expect(document.body).toMatchSnapshot();
        });
    });

    describe("on click, the sdk", () => {
        afterEach(() => {
            document.getElementsByTagName("html")[0].innerHTML = "";
        });

        test("should do nothing if data-cslp not available", async () => {
            const h1 = document.createElement("h1");

            document.body.appendChild(h1);
            new VisualEditor();

            h1.click();
            await sleep(0);

            expect(document.body).toMatchSnapshot();
        });

        describe("inline elements must be contenteditable", () => {
            test("single line should be contenteditable", async () => {
                const h1 = document.createElement("h1");

                h1.setAttribute(
                    "data-cslp",
                    "all_fields.blt58a50b4cebae75c5.en-us.single_line"
                );

                document.body.appendChild(h1);
                new VisualEditor();

                await sleep(0);
                h1.click();
                await sleep(0);

                expect(h1).toMatchSnapshot();
                expect(h1.getAttribute("contenteditable")).toBe("true");
            });

            test("multi line should be contenteditable", async () => {
                const h1 = document.createElement("h1");
                h1.setAttribute(
                    "data-cslp",
                    "all_fields.blt58a50b4cebae75c5.en-us.multi_line"
                );
                document.body.appendChild(h1);
                new VisualEditor();

                h1.click();
                await sleep(0);

                // h1.addEventListener("keydown", (e: KeyboardEvent) => {
                //     e.code.includes("")
                // })

                expect(h1).toMatchSnapshot();
                expect(h1.getAttribute("contenteditable")).toBe("true");
            });

            test("file should render a replacer and remove when it is not", async () => {
                const h1 = document.createElement("h1");
                h1.setAttribute(
                    "data-cslp",
                    "all_fields.blt58a50b4cebae75c5.en-us.file"
                );
                document.body.appendChild(h1);
                new VisualEditor();
                h1.click();
                await sleep(0);

                let replaceBtn = document.getElementsByClassName(
                    "visual-editor__replace-button"
                )[0];

                expect(h1).toMatchSnapshot();
                expect(replaceBtn).toMatchSnapshot();

                const h2 = document.createElement("h2");
                h2.setAttribute(
                    "data-cslp",
                    "all_fields.blt58a50b4cebae75c5.en-us.title"
                );
                document.body.appendChild(h2);

                h2.click();
                await sleep(0);

                replaceBtn = document.getElementsByClassName(
                    "visual-editor__replace-button"
                )[0];

                expect(replaceBtn).toBeUndefined();
            });
        });
    });
});

describe("visual editor DOM", () => {
    let h1: HTMLHeadElement;

    beforeAll(() => {
        FieldSchemaMap.setFieldSchema(
            "all_fields",
            getFieldSchemaMap().all_fields
        );
    });

    beforeEach(() => {
        h1 = document.createElement("h1");

        h1.setAttribute(
            "data-cslp",
            "all_fields.blt58a50b4cebae75c5.en-us.title"
        );

        h1.getBoundingClientRect = jest.fn(() => ({
            left: 10,
            right: 20,
            top: 10,
            bottom: 20,
        })) as any;

        document.body.appendChild(h1);
    });

    afterEach(() => {
        document.getElementsByTagName("html")[0].innerHTML = "";
    });

    afterAll(() => {
        FieldSchemaMap.clear();
    });

    test("should have an overlay over the element", async () => {
        new VisualEditor();

        let visualEditorOverlayWrapper = document.querySelector(
            `[data-testid="visual-editor__overlay__wrapper"]`
        );

        expect(visualEditorOverlayWrapper).toMatchSnapshot();

        await sleep(0);
        h1.click();
        await sleep(0);

        visualEditorOverlayWrapper = document.querySelector(
            `[data-testid="visual-editor__overlay__wrapper"]`
        );

        expect(visualEditorOverlayWrapper).toMatchSnapshot();
        expect(visualEditorOverlayWrapper?.classList.contains("visible")).toBe(
            true
        );

        const visualEditorWrapperTopOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--top"]`
        ) as HTMLDivElement;
        const visualEditorWrapperLeftOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--left"]`
        ) as HTMLDivElement;
        const visualEditorWrapperRightOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--right"]`
        ) as HTMLDivElement;
        const visualEditorWrapperBottomOverlay = document.querySelector(
            `[data-testid="visual-editor__overlay--bottom"]`
        ) as HTMLDivElement;

        expect(visualEditorWrapperTopOverlay.style.top).toBe("0px");
        expect(visualEditorWrapperTopOverlay.style.left).toBe("0px");
        expect(visualEditorWrapperTopOverlay.style.width).toBe("100%");
        expect(visualEditorWrapperTopOverlay.style.height).toBe(
            "calc(10px - 2px)"
        );

        expect(visualEditorWrapperBottomOverlay.style.top).toBe(
            "calc(20px + 2px)"
        );
        expect(visualEditorWrapperBottomOverlay.style.left).toBe("0px");
        expect(visualEditorWrapperBottomOverlay.style.width).toBe("100%");
        expect(visualEditorWrapperBottomOverlay.style.height).toBe(
            "calc(-20px - 2px)"
        );

        expect(visualEditorWrapperLeftOverlay.style.top).toBe(
            "calc(10px - 2px)"
        );
        expect(visualEditorWrapperLeftOverlay.style.left).toBe("0px");
        expect(visualEditorWrapperLeftOverlay.style.width).toBe(
            "calc(10px - 2px)"
        );

        expect(visualEditorWrapperRightOverlay.style.top).toBe(
            "calc(10px - 2px)"
        );
        expect(visualEditorWrapperRightOverlay.style.left).toBe(
            "calc(20px + 2px)"
        );
        expect(visualEditorWrapperRightOverlay.style.width).toBe(
            "calc(1004px - 2px)"
        );
    });

    test("should remove the DOM when method is triggered", async () => {
        const visualEditor = new VisualEditor();

        await sleep(0);
        h1.click();
        // We need this sleep as there are some async task happening while
        // the overlay is being rendered.
        await sleep(0);

        let visualEditorContainer = document.querySelector(
            `[data-testid="visual-editor__container"]`
        );

        expect(visualEditorContainer).toBeDefined();

        visualEditor.removeVisualEditorDOM();

        visualEditorContainer = document.querySelector(
            `[data-testid="visual-editor__container"]`
        );

        expect(visualEditorContainer).toBeNull();
    });

    test("should hide the DOM, when it is clicked", async () => {
        new VisualEditor();

        await sleep(0);
        h1.click();
        // We need this sleep as there are some async task happening while
        // the overlay is being rendered.
        await sleep(0);

        let visualEditorOverlayWrapper = document.querySelector(
            `[data-testid="visual-editor__overlay__wrapper"]`
        );
        expect(visualEditorOverlayWrapper?.classList.contains("visible")).toBe(
            true
        );
        expect(h1.getAttribute("contenteditable")).toBe("true");

        const visualEditorOverlayTop = document.querySelector(`
        [data-testid="visual-editor__overlay--top"]`) as HTMLDivElement;

        visualEditorOverlayTop?.click();

        visualEditorOverlayWrapper = document.querySelector(
            `[data-testid="visual-editor__overlay__wrapper"]`
        );
        expect(visualEditorOverlayWrapper?.classList.contains("visible")).toBe(
            false
        );
        expect(h1.getAttribute("contenteditable")).toBeNull();
    });
});
