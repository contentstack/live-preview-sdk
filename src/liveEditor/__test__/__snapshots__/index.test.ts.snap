// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Visual editor inline editing should add overlay to DOM when clicked 1`] = `
<body>
  <h1
    contenteditable="true"
    data-cslp="all_fields.blt58a50b4cebae75c5.en-us.modular_blocks.0.block.single_line"
    data-cslp-field-type="singleline"
  />
  <div
    class="visual-editor__container"
    data-testid="visual-editor__container"
  >
    <div
      class="visual-editor__cursor"
      data-testid="visual-editor__cursor"
    />
    <div
      class="visual-editor__overlay__wrapper visible"
      data-testid="visual-editor__overlay__wrapper"
    >
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--top"
        data-testid="visual-editor__overlay--top"
        style="top: 0px; left: 0px; width: 100%; height: 0px;"
      />
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--left"
        data-testid="visual-editor__overlay--left"
        style="left: 0px; top: 0px; height: 0px; width: 0px;"
      />
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--right"
        data-testid="visual-editor__overlay--right"
        style="left: 0px; top: 0px; height: 0px; width: 1024px;"
      />
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--bottom"
        data-testid="visual-editor__overlay--bottom"
        style="top: 0px; height: 0px; left: 0px; width: 100%;"
      />
      
        
      <div
        class="visual-editor__overlay--outline"
        data-testid="visual-editor__overlay--outline"
        style="top: 0px; height: 0px; width: 0px; left: 0px;"
      />
      
    
    </div>
    <button
      class="visual-editor__start-editing-btn"
      data-cslp-app-host="https://app.contentstack.com:443"
      data-cslp-branch="main"
      data-cslp-environment=""
      data-cslp-locale="en-us"
      data-cslp-stack=""
      data-testid="vcms-start-editing-btn"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        

        <g
          id="Edit"
        >
          

          <path
            clip-rule="evenodd"
            d="M3.58347 15.3803C3.35617 15.6076 3.22019 15.9104 3.20131 16.2313L3.00244 19.6122C2.95629 20.3967 3.60524 21.0456 4.38975 20.9995L7.7706 20.8006C8.0915 20.7817 8.39431 20.6458 8.62161 20.4185L20.6176 8.4225C21.1301 7.90993 21.1301 7.07891 20.6176 6.56634L17.4356 3.38436C16.923 2.8718 16.092 2.8718 15.5794 3.38436L3.58347 15.3803ZM4.32437 16.2974C4.32707 16.2515 4.3465 16.2083 4.37897 16.1758L14.2003 6.35446L17.4954 9.64949C17.5492 9.70337 17.6113 9.74403 17.6776 9.77148L7.82611 19.623C7.79364 19.6554 7.75038 19.6749 7.70454 19.6776L4.32369 19.8764C4.21161 19.883 4.11891 19.7903 4.1255 19.6782L4.32437 16.2974ZM18.4128 9.03624L19.8221 7.627C19.8953 7.55378 19.8953 7.43506 19.8221 7.36184L16.6401 4.17986C16.5669 4.10663 16.4481 4.10663 16.3749 4.17986L14.9958 5.55897L18.2908 8.854C18.3447 8.90788 18.3854 8.96996 18.4128 9.03624Z"
            fill="white"
            fill-rule="evenodd"
            id="Edit_2"
          />
          

        </g>
        

      </svg>
      

      <span>
        Start Editing
      </span>
    </button>
  </div>
</body>
`;

exports[`Visual editor on click, the sdk inline elements must be contenteditable file should render a replacer and remove when it is not 1`] = `
<h1
  data-cslp="all_fields.blt58a50b4cebae75c5.en-us.file"
  data-cslp-field-type="file"
/>
`;

exports[`Visual editor on click, the sdk inline elements must be contenteditable file should render a replacer and remove when it is not 2`] = `undefined`;

exports[`Visual editor on click, the sdk inline elements must be contenteditable multi line should be contenteditable 1`] = `
<h1
  contenteditable="true"
  data-cslp="all_fields.blt58a50b4cebae75c5.en-us.multi_line"
  data-cslp-field-type="multiline"
/>
`;

exports[`Visual editor on click, the sdk inline elements must be contenteditable single line should be contenteditable 1`] = `
<h1
  contenteditable="true"
  data-cslp="all_fields.blt58a50b4cebae75c5.en-us.single_line"
  data-cslp-field-type="singleline"
/>
`;

exports[`Visual editor on click, the sdk should do nothing if data-cslp not available 1`] = `
<body>
  <h1 />
  <div
    class="visual-editor__container"
    data-testid="visual-editor__container"
  >
    <div
      class="visual-editor__cursor"
      data-testid="visual-editor__cursor"
    />
    <div
      class="visual-editor__overlay__wrapper"
      data-testid="visual-editor__overlay__wrapper"
    >
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--top"
        data-testid="visual-editor__overlay--top"
      />
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--left"
        data-testid="visual-editor__overlay--left"
      />
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--right"
        data-testid="visual-editor__overlay--right"
      />
      
        
      <div
        class="visual-editor__overlay visual-editor__overlay--bottom"
        data-testid="visual-editor__overlay--bottom"
      />
      
        
      <div
        class="visual-editor__overlay--outline"
        data-testid="visual-editor__overlay--outline"
      />
      
    
    </div>
    <button
      class="visual-editor__start-editing-btn"
      data-cslp-app-host="https://app.contentstack.com:443"
      data-cslp-branch="main"
      data-cslp-environment=""
      data-cslp-locale="en-us"
      data-cslp-stack=""
      data-testid="vcms-start-editing-btn"
    >
      <svg
        fill="none"
        height="24"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        

        <g
          id="Edit"
        >
          

          <path
            clip-rule="evenodd"
            d="M3.58347 15.3803C3.35617 15.6076 3.22019 15.9104 3.20131 16.2313L3.00244 19.6122C2.95629 20.3967 3.60524 21.0456 4.38975 20.9995L7.7706 20.8006C8.0915 20.7817 8.39431 20.6458 8.62161 20.4185L20.6176 8.4225C21.1301 7.90993 21.1301 7.07891 20.6176 6.56634L17.4356 3.38436C16.923 2.8718 16.092 2.8718 15.5794 3.38436L3.58347 15.3803ZM4.32437 16.2974C4.32707 16.2515 4.3465 16.2083 4.37897 16.1758L14.2003 6.35446L17.4954 9.64949C17.5492 9.70337 17.6113 9.74403 17.6776 9.77148L7.82611 19.623C7.79364 19.6554 7.75038 19.6749 7.70454 19.6776L4.32369 19.8764C4.21161 19.883 4.11891 19.7903 4.1255 19.6782L4.32437 16.2974ZM18.4128 9.03624L19.8221 7.627C19.8953 7.55378 19.8953 7.43506 19.8221 7.36184L16.6401 4.17986C16.5669 4.10663 16.4481 4.10663 16.3749 4.17986L14.9958 5.55897L18.2908 8.854C18.3447 8.90788 18.3854 8.96996 18.4128 9.03624Z"
            fill="white"
            fill-rule="evenodd"
            id="Edit_2"
          />
          

        </g>
        

      </svg>
      

      <span>
        Start Editing
      </span>
    </button>
  </div>
</body>
`;

exports[`Visual editor should append a visual editor container to the DOM 1`] = `
<div
  class="visual-editor__container"
  data-testid="visual-editor__container"
>
  <div
    class="visual-editor__cursor"
    data-testid="visual-editor__cursor"
  />
  <div
    class="visual-editor__overlay__wrapper"
    data-testid="visual-editor__overlay__wrapper"
  >
    
        
    <div
      class="visual-editor__overlay visual-editor__overlay--top"
      data-testid="visual-editor__overlay--top"
    />
    
        
    <div
      class="visual-editor__overlay visual-editor__overlay--left"
      data-testid="visual-editor__overlay--left"
    />
    
        
    <div
      class="visual-editor__overlay visual-editor__overlay--right"
      data-testid="visual-editor__overlay--right"
    />
    
        
    <div
      class="visual-editor__overlay visual-editor__overlay--bottom"
      data-testid="visual-editor__overlay--bottom"
    />
    
        
    <div
      class="visual-editor__overlay--outline"
      data-testid="visual-editor__overlay--outline"
    />
    
    
  </div>
  <button
    class="visual-editor__start-editing-btn"
    data-cslp-app-host="https://app.contentstack.com:443"
    data-cslp-branch="main"
    data-cslp-environment=""
    data-cslp-locale="en-us"
    data-cslp-stack=""
    data-testid="vcms-start-editing-btn"
  >
    <svg
      fill="none"
      height="24"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      

      <g
        id="Edit"
      >
        

        <path
          clip-rule="evenodd"
          d="M3.58347 15.3803C3.35617 15.6076 3.22019 15.9104 3.20131 16.2313L3.00244 19.6122C2.95629 20.3967 3.60524 21.0456 4.38975 20.9995L7.7706 20.8006C8.0915 20.7817 8.39431 20.6458 8.62161 20.4185L20.6176 8.4225C21.1301 7.90993 21.1301 7.07891 20.6176 6.56634L17.4356 3.38436C16.923 2.8718 16.092 2.8718 15.5794 3.38436L3.58347 15.3803ZM4.32437 16.2974C4.32707 16.2515 4.3465 16.2083 4.37897 16.1758L14.2003 6.35446L17.4954 9.64949C17.5492 9.70337 17.6113 9.74403 17.6776 9.77148L7.82611 19.623C7.79364 19.6554 7.75038 19.6749 7.70454 19.6776L4.32369 19.8764C4.21161 19.883 4.11891 19.7903 4.1255 19.6782L4.32437 16.2974ZM18.4128 9.03624L19.8221 7.627C19.8953 7.55378 19.8953 7.43506 19.8221 7.36184L16.6401 4.17986C16.5669 4.10663 16.4481 4.10663 16.3749 4.17986L14.9958 5.55897L18.2908 8.854C18.3447 8.90788 18.3854 8.96996 18.4128 9.03624Z"
          fill="white"
          fill-rule="evenodd"
          id="Edit_2"
        />
        

      </g>
      

    </svg>
    

    <span>
      Start Editing
    </span>
  </button>
</div>
`;

exports[`visual editor DOM should have an overlay over the element 1`] = `
<div
  class="visual-editor__overlay__wrapper"
  data-testid="visual-editor__overlay__wrapper"
>
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--top"
    data-testid="visual-editor__overlay--top"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--left"
    data-testid="visual-editor__overlay--left"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--right"
    data-testid="visual-editor__overlay--right"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--bottom"
    data-testid="visual-editor__overlay--bottom"
  />
  
        
  <div
    class="visual-editor__overlay--outline"
    data-testid="visual-editor__overlay--outline"
  />
  
    
</div>
`;

exports[`visual editor DOM should have an overlay over the element 2`] = `
<div
  class="visual-editor__overlay__wrapper visible"
  data-testid="visual-editor__overlay__wrapper"
>
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--top"
    data-testid="visual-editor__overlay--top"
    style="top: 0px; left: 0px; width: 100%; height: 10px;"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--left"
    data-testid="visual-editor__overlay--left"
    style="left: 0px; top: 10px; width: 10px;"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--right"
    data-testid="visual-editor__overlay--right"
    style="left: 20px; top: 10px; width: 1004px;"
  />
  
        
  <div
    class="visual-editor__overlay visual-editor__overlay--bottom"
    data-testid="visual-editor__overlay--bottom"
    style="top: 20px; height: -20px; left: 0px; width: 100%;"
  />
  
        
  <div
    class="visual-editor__overlay--outline"
    data-testid="visual-editor__overlay--outline"
    style="top: 10px; left: 10px;"
  />
  
    
</div>
`;
